handler/exercism.lua

```diff
@@ -83,50 +42,97 @@ local function exercism_output_handler(options)
+        local indent = test_code:match('^%s+')
+        test_code = test_code:gsub('^' .. indent, ''):gsub('\n' .. indent, '\n'):gsub('%s+$', '')
+
+        result.tests[index] = {
+            name = element.name,
+            test_code = test_code
+        }
+
+        return nil, true
+    end
+
+    -- After a test has been executed, this handler is called, regardless of
+    -- whether it was successful or not. If a test has passed we add the status
+    -- to the current result.tests table. Failed or errored tests are handled
+    -- in seperate handlers below, which are called before this one.
+    handler.exercism_test_end = function(element, parent, status, debug)
+        if status == 'success' then
+            result.tests[index].status = 'pass'
         end
+
+        index = index + 1
+
+        return nil, true
     end
 
-    handler.suite_end = function()
+    -- If a test failed, this handler is called before the test_end handler.
+    -- We add the status and the failed message to the result.tests table.
+    local test_message_patt = '^.-%.lua:%d+:%s(.*)$'
+    handler.exercism_test_failure = function(element, parent, msg, debug)
+        result.tests[index].status = 'fail'
+        result.tests[index].message = msg:match(test_message_patt)
+
+        return nil, true
+    end
+
+    -- If a test errors, this handler is called before the test_end handler.
+    -- We add the status and the error message to the result.tests table.
+    handler.exercism_test_error = function(element, parent, msg, debug)
+        result.tests[index].status = 'error'
+        result.tests[index].message = msg
+
+        return nil, true
+    end
+
+    -- After a test suite ended, this handler is called. We evalute the
+    -- counters to set the overall status and then the result is written to
+    -- stdout as JSON.
+    handler.exercism_suite_end = function()
         if handler.errorsCount > 0 and handler.successesCount == 0 and handler.failuresCount == 0 then
             result.status = 'error'
             result.message = handler.errors[1].message
             result.tests = nil
         else
             result.status = handler.failuresCount == 0 and handler.errorsCount == 0 and 'pass' or 'fail'
-
-            add_to_tests(handler.successes, 'pass')
-            add_to_tests(handler.failures, 'fail')
-            add_to_tests(handler.errors, 'error')
         end
 
         io.write(json.encode(result))
     end
 
-    busted.subscribe({'suite', 'end'}, handler.suite_end)
+    -- Subscribing the handler functions to the different busted contexts.
+    busted.subscribe({'test', 'start'}, handler.exercism_test_start)
+    busted.subscribe({'test', 'end'}, handler.exercism_test_end)
+    busted.subscribe({'error', 'it'}, handler.exercism_test_error)
+    busted.subscribe({'failure', 'it'}, handler.exercism_test_failure)
+    busted.subscribe({'suite', 'end'}, handler.exercism_suite_end)
```

Comment:
I had no idea `busted` supported this (and I obviously didn't look too closely before). This is slick!